Heterogeneous architecture clusters (#1014)
openshift/enhancements: enhancements/multi-arch/heterogeneous-architecture.md

payload would be a manifestlist(containing a manifest for each architecture possessed by the CVO manifestlist). In either case, oc adm release new would permit non-CVO component images to be manifest or manifestlists and pass them through directly to the resultant release manifest(s). If a manifestlist release payload is generated, each architecture specific
release payload manifest will reference the same pullspecs provided in the input imagestream.

Since the internal Openshift registry does not support manifestlists, the release controller would use quay.io. There will be a new architecture option added to Cincinnati called "multi".

### Upgrades - Cincinnati and CVO

In addition to the homogenous payload graph that we have today, [support][OCPPLAN-7642] needs to be added for an additional graph for heterogeneous payloads which CVO will need to look at to make upgrade recommendations. There are a few upgrade scenarios to consider with the addition of heterogeneous payloads:

- heterogeneous -> heterogeneous payload upgrade
- homogenous -> heterogeneous payload upgrade

> - heterogeneous -> homogenous payload upgrade

Author: yeah , there is really no reason a user would need to go back to homogeneous except if they want to cut down on mirroring cost in the case of disconnected install when they decide to remove the node with the non control plane arch. I talked to Duncan and he felt the same so, I guess we can remove this.

> - heterogeneous -> homogenous payload upgrade

Author: since the plan is to not support CI for heterogeneous, the only question is pre-flight upgrade. can we just document that when you upgrade back to homogeneous to make sure the nodepools are not heterogeneous?

> - heterogeneous -> homogenous payload upgrade

Author: resolved after latest discussion.

The support for homogeneous payloads will exist so customers running disconnected homogeneous clusters do not incur the cost of mirroring the additional architectures if they were to use heterogeneous payloads. This means that in order for CVO to find the current release in the graph, it will have to know about the existing homogeneous graph as well in addition
to the new heterogeneous graph. There are few options on the best way to make update recommendations especially in the case of upgrading from a homogeneous to a heterogeneous payload:

- Have the heterogeneous graph include the homogeneous releases so CVO can make recommendations based on the current homogeneous release image. This is complicated by the fact that there will be multiple releases with the same version but different payloads. The trigger to update to a heterogeneous payload would be an option provided by the administrator through
either the ClusterVersion or Infrastructure API to "move" to the heterogeneous graph.

> - Have both the graphs be completely independent of each other and have an option in one of the aforementioned APIs to make the cluster heterogeneous and CVO would request the heterogeneous graph, find the current release version and update to it.

Reviewer: I was considering a mix of the two:

`oc adm release upgrade --multi-arch -to=<target>`
If `--multi-arch` is specified (or the cluster is already multi-arch), the Cincinnati query will be limited to the `multi` arch graph. If the _current_ cluster version does not exist in the `multi` arch graph with a connection to the target, then the upgrade is not permitted. If the current cluster is homogeneous, the customer will need to continue on in the homogeneous graph until their installed version is represented in the `multi` arch graph.
If the current and target version exist in the multi arch graph, and an upgrade occurs, then the clusterversion is updated with a net-new field to indicate that the cluster is now multi-arch. All subsequent queries will be for multi arch graphs (and ignore whether --multi-arch is specified).

As suggested in a previous comment, this would be a one-way street with no return to homogeneous outside of working with the support organization and forced upgrades.

- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for tech preview and GA
- [ ] User-facing documentation is created in [Openshift-docs](https://github.com/Openshift/Openshift-docs/)

## Summary

This enhancement describes the [plan][OCPPLAN-4577] to support provisioning and upgrades of heterogeneous architecture Openshift clusters. A heterogeneous cluster is a cluster with nodes running on different architectures with a homogeneous control plane and worker node pools supporting any combination of architectures (eg: x86 control plane + ARM worker nodes).
The node pools within the cluster are homogeneous. This document describes the support needed for heterogeneous architecture clusters in phases. There will be three major phases with basic support in phase 1 which is targeted for 4.11 and the other phases which address gaps beyond the 4.11 time frame. The targeted platforms in the initial phases are Azure
and AWS with support for x86 and ARM architectures.

![alt heterogeneous cluster](heterogeneous-cluster.png)

## Motivation

> The motivation for supporting heterogeneous architecture clusters is so users can take advantage of cost savings on architectures like ARM.

Reviewer 2: another motivation is to be able to colocate components that require different architectures, within the same cluster (avoiding cross-cluster latency and failure modes)

Author: added.

release payload manifest will reference the same pullspecs provided in the input imagestream.

Since the internal Openshift registry does not support manifestlists, the release controller would use quay.io. There will be a new architecture option added to Cincinnati called "multi".

### Upgrades - Cincinnati and CVO

In addition to the homogeneous payload graph that we have today, [support][OCPPLAN-7642] needs to be added for an additional graph for heterogeneous payloads which CVO will need to look at to make upgrade recommendations. There are a few upgrade scenarios to consider with the addition of heterogeneous payloads:

- heterogeneous -> heterogeneous payload upgrade
- homogeneous -> heterogeneous payload upgrade
- heterogeneous -> homogeneous payload upgrade

The support for homogeneous payloads will exist so customers running disconnected homogeneous clusters do not incur the cost of mirroring the additional architectures if they were to use heterogeneous payloads. This means that in order for CVO to find the current release in the graph, it will have to know about the existing homogeneous graph as well in addition

> to the new heterogeneous graph. There are few options on the best way to make update recommendations especially in the case of upgrading from a homogeneous to a heterogeneous payload:

Reviewer 2: i don't know how desirable it is(question for @sdodson ) but another approach is to just introduce an entirely new channel (e.g. stable-4.11-multiarch).

If you want to go from homogeneous to hetereogeneous, you would switch channels.

Author: resolved after recent discussions .

### Console

As of today console filters content based on control plane architecture. This needs to be entirely removed or made configurable for heterogeneous clusters.

### Install

For Phase 1, the way to add more workers to the cluster is to handcraft a machineset manifest file for a worker node and specify the architecture specific details like instance type, image references and deploy the machineset to get a worker node. This can be pseudo automated further by creating the manifests using `openshift-install create manifests` and modifying
or adding worker manifest files. Eventually the installer should support being able to declare differing compute and control plane sections in the install-config mainly allowing architectures to be different. In doing so it would also generate appropriate machineset and machineconfig files tailored to the specific architecture. The machineset file would contain
the instance types and region corresponding to the architecture chosen for the workers.

### Open Questions

- What would be the method to provide recommendation updates when moving from a homogeneous graph to a heterogeneous graph ?

> - Do OLM operator images have to include all architectures, even ones that are not supported by the operator ?

Reviewer 2: No.  If the publisher of an olm operator does not want to support a particular architecture, that is up to them.

(note that i'd expect all *redhat* operators to support all architectures that OCP supports, however).

This gets into the content filtering the console needs to do.  If i have a cluster that has ARM and x86 nodes:

* it's ok to install an operator that only supports ARM (there are workers where it can run)
* it's ok to install an operator that only supports x86 (there are workers where it can run)
* it's not ok to install an operator that only supports ppc + s390 (there will be nowhere it can run)
* sharp edge: it's not ok to install an operator that only supports ARM but expects to deploy a daemonset that runs on all workers.  I don't know that this scenario is likely enough to worry about for now.

So the console will need to be able to appropriately filter the set of olm operators that, if installed, will be able to find a worker to run on(meaning there is some overlap between the set of architectures the worker nodes are running on, and the set of architectures the olm operator claims to support)

Author: makes sense. updated the console section to reflect this and removed the question.

---
title: heterogeneous-architecture-clusters
authors:
  - "@Prashanth684"

> reviewers:

Reviewer 2: we will want the reviewer section to include every team lead of a team that is impacted by this (e.g. console, image registry, ART, etc) with a quick sentence explaining which piece of this impacts them (so they can focus their review on that section and not read the whole thing)

---
title: heterogeneous-architecture-clusters
authors:
  - "@Prashanth684"
reviewers:
  - "@bparees"
  - "@derekwaynecarr"
  - "@jupierce"
  - "@sdodson"
  - "@wking"
  - "@LalatenduMohanty"
  - "@sosiouxme"
  - "@enxebre"
  - "@dmage"
  - "@kdoberste"
  - "@adambkaplan"
  - "@perdasilva"

>   - "@mrunalp"

Reviewer 2: e.g. `@dmage - imagestream api multiarch changes`

>   - "@mrunalp"

Author: updated.

- Aimed at Azure/AWS with the primary supported architectures being x86 + ARM.
- For supporting ARM nodes on Azure, [hyper-v support][BZ-1949614] is required which is landing in RHEL 8.6 and will be included in RHCOS for 4.11.
- oc adm command changes to support manifestlist payloads.
- Ability to assemble and publish manifest listed release payloads in a reproducible way.
- Does not include upgrade support.
- Is not intended for production.

Phase 1: GA of basic support
- Release pipeline needs to be setup to construct manifestlist based release payloads.
- Upgrade support - CVO/Cincinnati support (heterogeneous -> heterogeneous payload upgrades only).
- Does not include support for:
  - image registry - internal image registry will not support manifest listed images.
  - imagestreams importing full manifestlist content (will continue to import a single image when referencing manifestlists).

Phase 2: Full basic functionality
- Imagestream manifestlist support.
- OLM/Operator ecosystem improvements.
- oc debug (and other commands) support for compute architecture.
- support toolbox.
- support for Power and Z architectures.

> - Hypershift awareness.

Reviewer 2: ```suggestion
- Hypershift support for heterogeneous management+worker clusters.
```

Phase 3: Quality of life improvements
- First class install support - ability to create non-control-plane-architecture workers at install time and generate machineconfig
and machineset manifests for the same.
- Improve console's content filtering by making it architecture aware.
- Possibly reduce mirroring costs by looking into things like sparse manifestlist.
- Upgrade support for homogeneous payload -> heterogeneous payload and vice versa.

### User Stories

- As an enterprise customer looking to save compute resource cost, I want to be able to run part of my application workload on ARM instead of x86.
- As the owner of an application that runs partially on x86 and partially on ARM, I want to be able to run the entire application within a
single cluster so I do not need to pay the cost of higher latency between clusters, or worry about the HA-resilience of cross-cluster bridges.

### API Extensions

Upgrades to a heterogeneous payload will require an additional field either in the Infrastructure or the ClusterVersion resource to indicate that

> the user wants to upgrade to a heterogeneous payload.

Reviewer 2: Will this get automatically set at install time when a user first installs a heterogeneous cluster?

what happens if it's not set on a heterogeneous cluster and an upgrade is attempted?

Author: i would expect CVO to automatically set it on first install based on whether its image is a manifestlist or not. Also on upgrades should CVO  also check the node architectures and automatically switch to heterogeneous payload if multiple arches are detected? i'll add that as an open question.

Reviewer 3: This API extension is only required once we get to the phase 3 support for conveniently swapping between homogeneous and heterogeneous release images.

Reviewer 4: Why can't we always use heterogeneous payload?

Author: for disconnected install cases we don't want the user to incur the additional cost of mirroring the extra arches if they only deal with homogeneous clusters.

Reviewer 2: for disconnected installs, it's a higher mirroring cost to mirror a bunch of images for architectures you will never use. 

(our mirroring tools will likely grow the ability to filter out images you don't want, in conjunction with registries allowing sparse manifestlists, someday, but that's not the world we are building this for today)

for connected clusters, i'm not sure there's a reason a user would care other than, perhaps, easier management of CVE exposures in cases of a CVE that impacts an architecture they aren't actually using.

Author: updated proposal to reflect the migration and upgrade aspects.

### Implementation Details/Notes/Constraints

Operators supporting multiple architectures cannot stop supporting a given architecture since it would break heterogeneous clusters which are
trying to upgrade.

### Risks and Mitigations

Heterogeneous payloads will only ship if the corresponding homogeneous payloads for all architectures can ship and are not broken due to a bug in a
specific architecture.

## Design Details

### RHCOS Support for ARM Azure

The RHCOS pipeline will need to generate Azure disk images for ARM. However as of today, RHEL does not support hyper-v for ARM and that support is
landing in RHEL8.6. Hence 4.11 builds of RHCOS for ARM Azure will be based on RHEL8.6. It is TBD whether a separate pipeline with early access RHCOS
builds will be needed as RHEL8.6 would not have GA'ed before 4.11.

### OCP Build & Release

> Phase 0 will involve building and publishing of [nightly builds][OCPPLAN-7640]. The release payload will be a manifest listed image with the component

Reviewer 2: ```suggestion
Phase 0 will involve building and publishing of [nightly builds][OCPPLAN-7640]. The release payload will be a manifestlist-based image with the component
```

> images being manifest listed as well. All 4 architectures - x86, ARM, Power and Z will be included in the manifestlist. Existing architecture specific

Reviewer 2: ```suggestion
images being manifestlist references as well. All 4 architectures - x86, ARM, Power and Z will be included in the manifestlist. Existing architecture specific
```

Reviewer 2: here it says all 4 arches will be included in phase 0, but earlier phase 0 was described as only having ARM+x86.

In either case, oc adm release new would permit non-CVO component images to be manifest or manifestlists and pass them through directly to the
resultant release manifest(s). If a manifestlist release payload is generated, each architecture specific release payload manifest will reference
the same pullspecs provided in the input imagestream.

Since the internal Openshift registry does not support manifestlists, the release controller would use quay.io. There will be a new architecture option
added to Cincinnati called "multi".

### Upgrades - Cincinnati and CVO

In addition to the homogeneous payload graph that we have today, [support][OCPPLAN-7642] needs to be added for an additional graph for heterogeneous
payloads which CVO will need to look at to make upgrade recommendations. There are a few upgrade scenarios to consider with the addition of
heterogeneous payloads:

- heterogeneous -> heterogeneous payload upgrade
- homogeneous -> heterogeneous payload upgrade
- heterogeneous -> homogeneous payload upgrade

The support for homogeneous payloads will exist so customers running disconnected homogeneous clusters do not incur the cost of mirroring the

> additional architectures if they were to use heterogeneous payloads. This means that in order for CVO to find the current release in the graph,it

Reviewer 2: ```suggestion
additional architectures if they were to use heterogeneous payloads. This means that in order for CVO to find the current release in the graph, it
```

will have to know about the existing homogeneous graph as well in addition to the new heterogeneous graph. There are few options on the best way
to make update recommendations especially in the case of upgrading from a homogeneous to a heterogeneous payload:

- Have the heterogeneous graph include the homogeneous releases so CVO can make recommendations based on the current homogeneous release image.
This is complicated by the fact that there will be multiple releases with the same version but different payloads. The trigger to update to a
heterogeneous payload would be an option provided by the administrator through either the ClusterVersion or Infrastructure API to "move" to the
heterogeneous graph.
- Have both the graphs be completely independent of each other and have an option in one of the aforementioned APIs to make the cluster heterogeneous

> and CVO would request the heterogeneous graph, find the current release version and update to it.

Reviewer 2: another option would be to have an entirely separate channel for heterogeneous payloads (e.g. `stable-multiarch-4.11`) and make admins switch channels if they want to move between heterogeneous and homogeneous payloads.

Author: i added this. so i guess the trigger for switching channels would also be the new field we introduce in the ClusterVersion or Infra resource to say "move to a heterogeneous" payload? or would admins have to specify the channel as well?

Reviewer 2: my idea was the admins would have to specify/change the channel explicitly.  not much point in having a new api field that modifies the subscribed channel.

Reviewer 2: i'd expect @sdodson and @wking to have the strongest opinions on whether a new multi-arch channel is appropriate/better/worse than the current api field proposal.

Reviewer 3: Currently channel selection encodes:

* A cap minor version, e.g. "I want to hear about updates to 4.9 and earlier, but not about updated into 4.10 and later".
* A hint at risk appetite for update-edge and target-release stability (`candidate`, `fast`, `stable`, and `eus`).

I'm not excited about mixing in architecture decisions there, although I agree that overloading channel like that might help avoid growing a new knob for heterogeneous vs. homogeneous.  With the channel approach, the transition would look like:

1. User is running version A on homogeneous `stable-4.11`.
2. User decides to move to heterogeneous releases.
3. User sets a `stable-multiarch-4.11` channel.
4. CVO fetches that graph.  Today we'd successfully find the current release and avoid `VersionNotFound` because the CVO [naively assumes that it can match on the version string, without worrying if the image digest matches][1].
5. User has to update within this graph to version B, or explicitly `oc adm upgrade --allow-explicit-upgrade --to-image "${BY_DIGEST_PULLSPEC_FOR_HETEROGENOUS_RELEASE_A}"` to make the transition.

With the arch approach, the flow is:

1. User is running version A on homogeneous `stable-4.11`.
2. User decides to move to heterogeneous releases.
3. User sets the property in Infrastructure or ClusterVersion or wherever.
4. CVO fetches that graph, finds the current release A with the new architecture, and initiates an "update" to roll it out.
5. If the user wants an update to release B as well, they can kick that off now completely within the heterogeneous graph.

[1]: https://github.com/openshift/cluster-version-operator/blob/3db74ebb012838ffdd776ca65210e4d6d66b7472/pkg/cincinnati/cincinnati.go#L127-L146

Reviewer 5: The decision to move to a multiarch should a configuration change for the user. The UX of only changing a channel does not fit in that model. I think this flow would work better from the UX point of view
```
With the arch approach, the flow is:

User is running version A on homogeneous stable-4.11.
User decides to move to heterogeneous releases.
User sets the property in Infrastructure or ClusterVersion or wherever.
CVO fetches that graph, finds the current release A with the new architecture, and initiates an "update" to roll it out.
If the user wants an update to release B as well, they can kick that off now completely within the heterogeneous graph
```

These options are under discussion and a finalized proposal will be put forth in the ticket for upgrades referenced above along with the API that would
change to include the option to indicate a move to the heterogeneous payload.

Cincinnati needs to [support][OCPPLAN-7643] parsing manifestlist images to be able to upgrade to a heterogeneous architecture payload.

### Imagestreams

Imagestreams need to better support manifestlist images. As of today they do recognize manifestlists, but it only imports images matching the control
plane architecture. The idea here would be to import the manifest listed image in its entirety. This also has other implications such as the fact that
the internal registry does not support manifest listed images. So, with full manifestlist support in imagestreams, pullthrough is not possible for
manifestlists. We would ask users to use quay.io rather than use the internal registry if full manifestlist support is needed.

The oc commands like debug, must-gather etc use the imagestreams and therefore would use an x86 image to deploy on the worker nodes which would fail

> with an ARM worker. As a workaround, they can be provided the manifest listed image with the `--image` parameter which would allow the commands to

Reviewer 2: ```suggestion
with an ARM worker. As a workaround until imagestreams support importing all the architectures, they can be provided the manifest listed image with the `--image` parameter which would allow the commands to
```

work on the worker nodes as well.

### OLM

Operators need to specify which architectures they support based on annotations. There also needs to be a mechanism to ensure that the architecture is
actually supported if that label is present. Also, all architectures should share the same content. Images potentially need to be available on

> architectures where there is no support for the operator as well. But the filtering should make sure only the supported architectures show up.

Reviewer 2: I'm not clear on why "Images potentially need to be available on
architectures where there is no support for the operator as well"  ?

Author: removed it based [on the previous comment](https://github.com/openshift/enhancements/pull/1014#discussion_r793106277) and added the fact that operators/operands need to be nodeselected.

---
title: heterogeneous-architecture-clusters
authors:
  - "@Prashanth684"
reviewers:
  - "@bparees"
  - "@derekwaynecarr"
  - "@jupierce"
  - "@sdodson"
  - "@wking"
  - "@LalatenduMohanty - upgrades"
  - "@sosiouxme - payload build and release"
  - "@enxebre - hypershift"
  - "@dmage - imagestream api changes"
  - "@kdoberst - console filtering changes"
  - "@adambkaplan - Openshift Builds"
  - "@perdasilva - OLM changes"
  - "@mrunalp - MCO changes"
approvers:
  - "@bparees"
  - "@derekwaynecarr"
creation-date: 2022-01-17
last-updated: 2022-01-17
status: implementable
---

# heterogeneous-architecture-cluster-support

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for tech preview and GA
- [ ] User-facing documentation is created in [Openshift-docs](https://github.com/Openshift/Openshift-docs/)

## Summary

> This enhancement describes the [plan][OCPPLAN-4577] to support provisioning and upgrades of heterogeneous architecture Openshift clusters.

Reviewer 3: nit: the OCPPLAN-4577 link, and/or a public analog like [RFE-2240](https://issues.redhat.com/browse/RFE-2240) can go in [the `tracking-link` property][1] in the YAML frontmatter.

[1]: https://github.com/openshift/enhancements/blame/66eb520cb9606170d293432ec3118b04dbfdfa69/guidelines/enhancement_template.md#L15

Author: linked. thanks for pointing this out!

A heterogeneous cluster is a cluster with nodes running on different architectures with a homogeneous control plane and worker node pools
supporting any combination of architectures (eg: x86 control plane + a mix of ARM/P/Z worker nodes). The node pools within the cluster are
homogeneous. This document describes the support needed for heterogeneous architecture clusters in phases. There will be three major phases

> with basic support in phase 1 which is targeted for 4.11 and the other phases which address gaps beyond the 4.11 time frame. The targeted

Author: clarified that it is techpreview.

platforms in the initial phases are Azure and AWS with support for x86 and ARM architectures.

![alt heterogeneous cluster](heterogeneous-cluster.png)

## Motivation

- Users can take advantage of cost savings on architectures like ARM.
- Users can colocate applications that require different architectures rather than needing multiple clusters and thus avoid cross cluster
latency and connectivity failures.

### Goals

> - Provide a way to install and upgrade clusters with varied architecture node pools with the ability to autoscale heterogeneous machinesets.

Author: i should change this to machineset. That was what was meant here. like you said nodepool is specific in the context of hypershift.

- Support provisioning heterogeneous architecture clusters in Hypershift.

> - Generate a manifestlist based payload that contains manifestlist references for the 4 architectures OpenShift supports (x86, ARM, Power and Z).

Reviewer 5: Is the goal to have different architectures among worker nodes as well? For example if there are 100 worker nodes, can be 25 nodes of x86, 25 nodes of ARM, 25 nodes of power?

Reviewer 2: yes....the goal is not just "control plane nodes are one arch, and workers are a different arch", the goal is "workers can have a mix of arches"

Author: changed the wording in the summary section to reflect this.

trying to upgrade.

### Risks and Mitigations

Heterogeneous payloads will only ship if the corresponding homogeneous payloads for all architectures can ship and are not broken due to a bug in a
specific architecture.

## Design Details

### RHCOS Support for ARM Azure

The RHCOS pipeline will need to generate Azure disk images for ARM. However as of today, RHEL does not support hyper-v for ARM and that support is
landing in RHEL8.6. Hence 4.11 builds of RHCOS for ARM Azure will be based on RHEL8.6. It is TBD whether a separate pipeline with early access RHCOS
builds will be needed as RHEL8.6 would not have GA'ed before 4.11.

### OCP Build & Release

Phase 0 will involve building and publishing of [nightly builds][OCPPLAN-7640]. The release payload will be a manifestlist-based image with the
component images being manifestlist references as well. All 4 architectures - x86, ARM, Power and Z will be included in the manifestlist.
Existing architecture specific release payloads continue to be delivered through their individual release controllers. A new heterogeneous

> release controller is bootstrapped and displays "4-stable" entries for ART constructed release payloads. `oc adm release ...` verbs would be

Reviewer 3: The release controller is up at https://multi.ocp.releases.ci.openshift.org/ , if we want to link that here.  It's just not getting regular nightly builds yet.

Author: linked it below in the infrastructure section.

[enhanced][WRKLDS-370] to allow the creation and management of manifestlist release payloads. Sub-verbs like `mirror`, `info`, and `extract`
must be adapted to support manifestlist based payloads. The trigger for creating the heterogeneous payloads will be whenever there is a new
payload produced for one of the included architectures.

In the case of `oc adm release new`, the manifestlist flow would be triggered whenever the CVO image in a source image/imagestream is a manifestlist.
- If the source CVO image is a standard manifest, the generated release payload will also be a manifest.
- If the source CVO image is a manifestlist, the generated release payload would be a manifestlist(containing a manifest for each architecture
possessed by the CVO manifestlist).

In either case, oc adm release new would permit non-CVO component images to be manifest or manifestlists and pass them through directly to the
resultant release manifest(s). If a manifestlist release payload is generated, each architecture specific release payload manifest will reference

> the same pullspecs provided in the input imagestream.

Reviewer 3: is it worth any guards here around weird architecture mixing.  For example, if the CVO ref happens to be amd64, but all the other referenced images are s390x, should `oc` complain about that?  Or just silently assume the ImageStream admins know what they're doing and package it without complaint?

Ah, also, release images lean on the ImageStream format today, so we'd need at least some ImageStream handling that didn't care if the referenced releases were manifest lists or not for phase0, even if that isn't "the in-cluster ImageStream controller can handle manifest lists".

Author: the non CVO image arch references should be a subset of the CVO image arch references. I think oc should complain about this scenario and reject creating the payload.

Since the internal Openshift registry does not support manifestlists, the release controller would use quay.io. There will be a new architecture option

> added to Cincinnati called "multi".

Reviewer 3: This `Cincinnati` sentence seems like it should be shifted down to the next section.

Author: done.

- Have an independent channel for heterogeneous payloads and have administrators switch channels if they want to move between homogeneous and
heterogeneous payloads

These options are under discussion and a finalized proposal will be put forth in the ticket for upgrades referenced above along with the API that would
change to include the option to indicate a move to the heterogeneous payload.

Cincinnati needs to [support][OCPPLAN-7643] parsing manifestlist images to be able to upgrade to a heterogeneous architecture payload.

### Imagestreams

Imagestreams need to better support manifestlist images. As of today they do recognize manifestlists, but it only imports images matching the control
plane architecture. The idea here would be to import the manifest listed image in its entirety. This also has other implications such as the fact that
the internal registry does not support manifest listed images. So, with full manifestlist support in imagestreams, pullthrough is not possible for
manifestlists. We would ask users to use quay.io rather than use the internal registry if full manifestlist support is needed.

> The oc commands like debug, must-gather etc use the imagestreams and therefore would use an x86 image to deploy on the worker nodes which would fail

Reviewer 4: Can we just change these image streams (must-gather, etc) to use `reference: true`? I don't think we really need image stream functionality (ability to trigger new deployments, etc.) for these images.

```
spec:
  tags:
  - from:
      kind: DockerImage
      name: registry...@sha256:...
    name: latest
    reference: true
```

Reviewer 2: remind me, does that effectively mean the imagestreamtag SHA will just be the manifestlist SHA?

Reviewer 4: from.name will be copied as is into status:

```
status:
  tags:
  - items:
    - created: "2022-02-02T19:01:17Z"
      dockerImageReference: registry...@sha256:...
      generation: 3
      image: ""
    tag: latest
```

With such a change:
```
$ oc get -n openshift istag must-gather:latest
Error from server (NotFound): imagestreamtags.image.openshift.io "must-gather:latest" not found
$ oc get -n openshift imagetag must-gather:latest
NAME                 SPEC        STATUS                   HISTORY   UPDATED
must-gather:latest   Scheduled   registry...@sha256:...   1         3 hours ago
```

Reviewer 2: i guess since from.name is already a SHA for those in-payload imagestreams, it works ok

Author: i just checked that adding the `reference: true`  tag to the tools imagestream works fine when i do an oc debug node on a hetero arch cluster for any of the nodes. so this is something that would have to be added in the imagestream manifests in cluster samples right? and they would add it only for a heterogeneous deployment? would it cause complications during upgrade to homogeneous ? or would we never use the internal registry for these images going forward?

Reviewer 4: it will make impossible to use pull-through for these images, i.e. you won't be able to pull it as `image-registry.openshift-image-registry.svc:5000/openshift/must-gather:latest`

Reviewer 4: CVO may need to learn about this flag: https://github.com/openshift/cluster-version-operator/blob/d3f4ca856badfe3db7a74415bd4a418d128999b1/lib/resourcemerge/imagestream.go#L32

with an ARM worker. As a workaround until imagestreams support importing all the architectures, they can be provided the manifest listed image with the
`--image` parameter which would allow the commands to work on the worker nodes as well.

### OLM

Operators need to specify which architectures they support based on annotations. There also needs to be a mechanism to ensure that the architecture is
actually supported if that label is present. Also, all architectures should share the same content.

OLM would need to make sure that it deploys operators only to nodes whose architecture matches the architectures on which the operator is supported.

> This would mean that the OLM would have to be told which architectures an operator supports by the Console operator. Today, OLM has a subscription

Reviewer 2: technically olm ought to know which architectures the operator supports.  but the admin installing the operator may not want to deploy it on all of those architectures.

so the olm piece of this is just updating the subscription api so that nodeaffinity can be specified.

the console piece will be giving the user a way to specify what that nodeaffinity should be, as part of the install (and defaulting it appropriately)

Author: updated.

  - references to some images like cluster-api-provider-aws, apiserver-network-proxy etc.. are x86. these would need to change to provide the images
  for other architectures too. These images can be overridden through annotations today.
- Another option would be to mandate an x86 worker node be present for Hypershift and nodeselect the x86 node for the hosted cluster control plane
deployment.
- When creating the nodepool, the defaults for instances types and AMIs are x86 specific. The instance types can be overridden but not the AMIs.
An option needs to be added to the create cluster command to override AMIs.

### Console

As of today console filters content based on control plane architecture. This needs to be entirely removed or made configurable for
heterogeneous clusters. The operator should be able to filter out the set of OLM operators that, if installed will be able to find
worker nodes to run on. This means that it has to display only operators that can be supported by architectures that the workers are
running on. Additionally, it would also have to tell OLM which architectures to nodeselect on by creating a subscription object and

> adding a podAffinity rule with nodeselectorterms. The support for podAffinity in the subscription object will be added to OLM.

Reviewer 2: user should be able to choose which architectures are put into the nodeaffinity on the subscription.  but the console should default it to whatever the operator supports (user may wish to deselect some architectures).

also, a nit, i'd make this a separate paragraph because they are two different features the console needs to build and people might gloss over it as it is currently written.

Author: updated.

### Install

For Phase 1, the way to add more workers to the cluster is to handcraft a machineset manifest file for a worker node and specify the architecture
specific details like instance type, image references and deploy the machineset to get a worker node. This can be pseudo automated further by creating
the manifests using `openshift-install create manifests` and modifying or adding worker manifest files. Eventually the installer should support being
able to declare differing compute and control plane sections in the install-config mainly allowing architectures to be different. In doing so it would
also generate appropriate machineset and machineconfig files tailored to the specific architecture. The machineset file would contain the instance
types and region corresponding to the architecture chosen for the workers.

> ### Open Questions

Reviewer 6: What is going to happen with the Samples Operator? This currently only installs imagestreams which match the control plane arch. Not all sample images have the same level of mutli-arch support.

cc @dperaza4dustbit

Reviewer 2: this is somewhat dependent on what happens with imagestreams themselves.  My ideal outcome is that imagestreams grow a new api field that says "import all manifests" (or maybe just "import the manifestlist, don't extract a manifest from within it, as it does today").

The samples imagestreams would then be updated to set that new api field.

At that point you at least have the ability to use any arch the sample makes available.  It doesn't solve the problem of a particular sample only publishing x86 images, however.  The next step is probably annotating the imagestreams in a way that makes it clear which arches they offer/support, similar to the way OLM operators are annotated.

Given we already maintain imagestream files on a per-arch basis today, determining which imagestreams support which arches should be trivial.

Author: how would samples know that it has to set that api field in its imagestreams ? i.e, would it have to know that it is deployed on a heterogeneous cluster as opposed to  homogeneous somehow ?

Reviewer 2: we'd always set it.  There's no harm I can think of in setting it on a homogeneous cluster.

Reviewer 2: The images should already be published as manifestlists, but the image streams created by the samples operator will need to be configured to import the content as a manifestlist instead of just picking one image from the manifestlist as it does today. How this is configured will depend on the image stream manifestlist api config support works.

And the image streams are maintained by the teams that own the images, not by the samples operator team, for the most part. But David Peraza should be able to help wrangle that.

Author: added a samples section capturing this.

- What would be the method to provide recommendation updates when moving from a homogeneous graph to a heterogeneous graph ?

### Test Plan

Testing for Phase 1:
- QE would need to test heterogeneous nightly payloads before the entry is added to Cincinnati. This implies that the heterogeneous payloads will be

> released a bit later than the homogeneous payloads.

Reviewer 3: "released" is a bit fuzzy.  We currently build all the homogeneous arch releases and then add them to the `candidate` channels.  But the release isn't GA until errata is published, usually a week later.  I don't think anyone will be particularly worried about a small delay before entry in the `candidate` channels.  Especially for a TechPreview phase 1.  But it might help make this clearer if you replace "released" with "pushed to Quay" or "added to `candidate` channels" or whatever point in the process you're focusing on here.

Author: yes this is true but the intent here was that we did not want to hold up homogeneous releases if the heterogeneous testing uncovered some issues which would make it  release blocking. i will change the wording here.

- New release payload acceptance jobs on heterogeneous clusters.

Testing for Phase 2:
- With the addition of Power and Z architectures, there will be more permutations to test. We would need to figure out the absolute essential set of
permutations and the tests to run on those.
- Imagestream conformance tests.
- OLM and console testing
- Test various upgrade scenarios (homogeneous -> heterogeneous and vice versa)

### Graduation Criteria

Phase 1 will have a pre-4.11 dev preview and 4.11 tech preview with documented limitations.

#### Dev Preview -> Tech Preview

There will be a tech preview of basic functionality (Phase 1 as described above) in 4.11

#### Tech Preview -> GA

TBD

#### Removing a deprecated feature

This enhancement does not describe removing a feature of Openshift, this section is not applicable.

### Upgrade / Downgrade Strategy

The upgrade strategy is described in the design details section above. The important changes here are having a separate architecture
for the heterogeneous releases and also the work needed to support upgrades from a homogeneous release -> heterogeneous release and
vice versa in the future.

### Version Skew Strategy

> TBD

Reviewer 3: Also somewhat in the version-skew front would be "updates" between homogeneous and heterogeneous releases.  Obviously if a cluster had attached heterogeneous compute, we'd want to try and block heterogeneous-to-homogeneous transitions to avoid breaking that compute.  But that's always going to be at least a bit racy, e.g. if someone launches their first alt-arch compute node after initiating the transition to a homogeneous release image but before the CVO had gotten deep enough in to move the machine-(config-)API portions over to homogeneous images.  What would happen to those machines?  Would the lock when the MCO tries to update them into a homogeneous (and incompatible) RHCOS?  And then maybe a MachineHealthCheck would trigger the machine-API components to gracefully reap the alt-arch nodes?  Probably not too many admins would attempt something like this, but it might be worth at least talking through the broad strokes of how we expect this to work out.

Author: the upgrade of machine-os-content on that node will fail, but i'm not sure if machine-api will reap the node. The other thing here is, if the alt-arch node comes up as schedulable (while on the previous heterogeneous release), some operators (for the homogeneous release) could be scheduled on that node causing them to fail and thus failing the upgrade in that state even before getting to the above mentioned scenario.

Author: mentioned this in version skew although not sure what can be dine here other than the admin scaling down that machineset manually and restarting the upgrade. slightly related - the machine api folks are also having a conversation on whether machinesets' instance type and image id should be allowed to be modified post installation.

There will be minor changes to the ClusterVersion or infrastructure APIs for supporting upgrades and possibly Hypershift. The imagestreams API
would also undergo minor changes to introduce support for manifestlist import behavior.

#### Failure Modes

No new failure modes are introduced as part of this feature.

#### Support Procedures

TBD

## Implementation History

A heterogeneous payload was built as a proof of concept and spun up worker nodes using ARM with an x86 control plane on AWS. It worked,
but here are the OCP features we know won’t work properly in such an environment today which are explained in detail above:
- Imagestreams - does not fully support manifestlists, they will import a single architecture (by default based on the control plane’s architecture,
though it can be chosen explicitly) from the manifestlist, so anything referencing an imagestream tag will get the image associated with that
architecture, regardless of what architecture the node is running.
- Internal registry - does not support manifestlists (storing or pulling). This also means you cannot use the internal registry for image
pull-through (proxying) for manifestlist-based content.

> - Builds - do not produce manifestlists. Builds do support use of nodeselectors to steer the build to a node of the appropriate architecture.

Reviewer 6: It is also challenging for builds to push manifest lists in a way that that is not susceptible to race conditions. Multi-arch build fan-outs is a very real use case for heterogeneous clusters.

cc @nalind @BillDett - perhaps we can discuss adding "Etag" support to Quay?

Reviewer 4: @adambkaplan what kind of race conditions?

Reviewer 2: yeah, pushing manifestlists is definitely something that needs to be part of a pipeline(run all builds, wait for them all to complete/push, then create the manifestlist), not a single build step.  Which is part of why the build v1 api is not going to grow support for pushing them.

it should of course be able to consume base images from manifestlists, but that should already work today.

Reviewer 7: If two nodes of different architectures, building images for their local architectures, both create manifest lists that list the image that they just built and then try to push those lists to a registry using the same tag, there's currently no facility for merging their results, so when the first builder pushes its list with one entry in it, it will succeed, and when the second builder pushes its list with one entry in it, it will overwrite the list that the first builder pushed, effectively removing the first builder's image from the list in the registry.

Reviewer 4: @adambkaplan so you want builders to append new architectures into the manifest list? The first builder will need to erase all other architectures as manifest lists are supposed to have images with the same version of the application. And this feels a bit wrong, especially if we talk about tags like `latest`. I mean, it's bad to miss architectures in a manifest list and/or to have images that are built from different commit-ishes, so I think @bparees's option is the only way to go.

Reviewer 8: I think you are all saying in a very roundabout way that there needs to be a controlling process for multi-architecture builds.  How could there *not* be?

Reviewer 8: So is there any basic code that accepts something that looks like an OCP Build objects but handles the arch steering and manifest list aggregation that isn't tied up with a lot of higher level stuff (i.e. isn't OSBS)?

## Summary

This enhancement describes the plan to support provisioning and upgrades of heterogeneous architecture Openshift clusters.
A heterogeneous cluster is a cluster with nodes running on different architectures with a homogeneous control plane and worker node pools
supporting any combination of architectures (eg: x86 control plane + a mix of ARM/P/Z worker nodes). The node pools within the cluster are
homogeneous. This document describes the support needed for heterogeneous architecture clusters in phases. There will be three major phases
with basic support in phase 1 which is targeted for 4.11 and the other phases which address gaps beyond the 4.11 time frame. The targeted
platforms in the initial phases are Azure and AWS with support for x86 and ARM architectures.

![alt heterogeneous cluster](heterogeneous-cluster.png)

## Motivation

- Users can take advantage of cost savings on architectures like ARM.
- Users can colocate applications that require different architectures rather than needing multiple clusters and thus avoid cross cluster
latency and connectivity failures.

### Goals

> - Provide a way to install and upgrade clusters with varied architecture node pools with the ability to autoscale heterogeneous machinesets.

Reviewer 9: Changes in the machine API to support auto scaling from zero are probably needed to make the autoscaler aware of the upcoming Node arch cc @JoelSpeed @elmiko

Reviewer 10: i'm not sure how the autoscaler handles the architecture of the workload in the scheduler. i'm guessing this is already part of kube's scheduler, so it probably "just works", but i'm guessing we will need to pass the cpu architecture in addition to cpu cores at the least. this will probably require a little investigation into how the autoscaler handles the architecture.

Reviewer 10: makes perfect sense, thanks for the guidance =)

Author: will create a machine-api section and put this in. thanks!

Author: added the new section for machine api changes. @JoelSpeed @elmiko let me know if there are any changes/addition needed.

and CVO would request the heterogeneous graph, find the current release version and update to it.
- Have an independent channel for heterogeneous payloads and have administrators switch channels if they want to move between homogeneous and
heterogeneous payloads

These options are under discussion and a finalized proposal will be put forth in the ticket for upgrades referenced above along with the API that would
change to include the option to indicate a move to the heterogeneous payload.

Cincinnati needs to [support][OCPPLAN-7643] parsing manifestlist images to be able to upgrade to a heterogeneous architecture payload. There will be a
new architecture option added to it called "multi".

### Imagestreams

Imagestreams need to better support manifestlist images. As of today they do recognize manifestlists, but it only imports images matching the control
plane architecture. The idea here would be to import the manifest listed image in its entirety. This also has other implications such as the fact that
the internal registry does not support manifest listed images. So, with full manifestlist support in imagestreams, pullthrough is not possible for

> manifestlists. We would ask users to use quay.io rather than use the internal registry if full manifestlist support is needed.

Reviewer 4: Our end-to-end tests rely on `image-registry.openshift-image-registry.svc:5000/openshift/tools:latest` (used in 53 files). It provides `oc` compatible with the current cluster. Will we migrate to something else?

Reviewer 2: depends on whether you're going to make pullthrough work for manifestlists for the internal registry :)

if not then i'd say we have to.

This may also have implications for the console's CLI downloader (i'm not sure what it points to today)

Author: the CLI downloader gets it from the downloads endpoint, so they're fine. And we have oc binaries for both arm and x86 in those.

object which is used to tell it how to deploy an operator. It has support for the nodeselector field but this would not be sufficient to specify
filtering on multiple architectures. To this end, it would have to support the podAffinity field which uses the nodeselectorterms to filter on
multiple values.

### Machine Config Operator

No changes are needed in the MCO for Phase 1 as the machine-os-content image would be a manifest listed image and the machine config daemon would
extract the relevant architecture's machine-os-content based on the node it runs on. In the future, architecture specific configurations should
not be required, instead they need to be templatized and be generalized for certain items, for example kubelet system reserved memory could scale
based on page size.

### Hypershift

The addition of heterogeneous support in Openshift also opens up a new dimension for [support][OCPPLAN-5684] in Hypershift where the control plane 
and the workers could run on different architectures. There would be certain changes needed to realize this as there are limitations which need to
be addressed:

> - When creating the hosted cluster, we need to override the `--release-image` option to specify the manifest listed release image in order

Reviewer 9: Unless there's a reason to not, I'd be multi-arch by default out of the box without requiring user explicit input, i.e we'd enforce that the multi-arch release image is used. Then if you create an arm NodePool it just works.

Reviewer 9: yeh I think we'll do have to support disconnected, we'l figure the best UX. Your statements here seems accurate to me, thanks.

> - When creating the hosted cluster, we need to override the `--release-image` option to specify the manifest listed release image in order

Reviewer 9: Actually can we please drop `--release-image` and refer to HostedCluser and NodePool APIs.
Our cli is currently a helper for dev purposes and it does not define HyperShift, the API does.

Author: ok got it. will do.

to be able to deploy on a worker node of any architecture. Optionally, the config could provide a flag to indicate that it is being installed
on a heterogeneous architecture cluster in which case it will pull down a manifest listed release payload rather than the x86 release payload.
- For the hosted control plane to work on non x86 architecture nodes, the following image references would need changing:

>   - the release image fixtures referencing x86 imagestreams. these would need to be changed to manifestlist references or there would need

Reviewer 9: Most of our Control plane components images are inferred from the spec.releaseImage so if that's multi arch the components will be.

Author: ok..so even images like cluster-api-provider here will be from the payload? https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go#L24

Reviewer 9: nope, that statement is correct, we'll need to make sure the ones which are statically pinned are multi-arch: CAPI, Hypershift itself and a few others

  to be imagestreams added for individual architectures and they would be picked based on the node architecture.
  - references to some images like cluster-api-provider-aws, apiserver-network-proxy etc.. are x86. these would need to change to provide the images
  for other architectures too. These images can be overridden through annotations today.

> - Another option would be to mandate an x86 worker node be present for Hypershift and nodeselect the x86 node for the hosted cluster control plane

Reviewer 9: >Another option would be to mandate an x86 worker node be present for Hypershift and nodeselect the x86 node for the hosted cluster control plane

I think this is a non-go, A Hosted Cluster might require multiple Nodes e.g HA and the coupling with the arch is unnecessary. We'll need to make sure statically pinned Images are multi-arch

Author: will remove this as an option.

deployment.

> - When creating the nodepool, the defaults for instances types and AMIs are x86 specific. The instance types can be overridden but not the AMIs.

Reviewer 9: The AMI it's a choice, it can be overridden https://github.com/openshift/hypershift/blob/61272973b6134cb6ef9702779912c2eb0aee76c2/api/v1alpha1/nodepool_types.go#L326

We happen to get the default as the x86_64 from the payload which will need to adapt
https://github.com/openshift/hypershift/blob/9b0c2afbbaee8e4330d8a388498d6023e7508c7f/hypershift-operator/controllers/nodepool/nodepool_controller.go#L993-L996

Author: done. please let me know if that suffices.

Author: @enxebre could you check and let me know if the Hypershift section addresses all of your concerns? thanks!

heterogeneous clusters. The operator should be able to filter out the set of OLM operators that, if installed will be able to find
worker nodes to run on. This means that it has to display only operators that can be supported by architectures that the workers are
running on.
- Console would also have to tell OLM for each operator which architectures to nodeselect on by creating a subscription object and adding a
podAffinity rule with nodeselectorterms. The nodeselectorterms will be defaulted with the architectures that the operator supports, but would
also have a provision for users to specify the architectures they want (from the set of supported architectures).

### Install

For Phase 1, the way to add more workers to the cluster is to handcraft a machineset manifest file for a worker node and specify the architecture
specific details like instance type, image references and deploy the machineset to get a worker node. This can be pseudo automated further by creating
the manifests using `openshift-install create manifests` and modifying or adding worker manifest files. Eventually the installer should support being

> able to declare differing compute and control plane sections in the install-config mainly allowing architectures to be different. In doing so it would

Reviewer 9: > The machineset file would contain the instance types and region

nit: The machineset file would contain the instancetype, AMI and region

also generate appropriate machineset and machineconfig files tailored to the specific architecture. The machineset file would contain the instance

> types and region corresponding to the architecture chosen for the workers.

Reviewer 9: Risk/UX flaw: if a user ever changes the the AMI and type of an existing MachineSet e.g x86 -> arm, upcoming Machines will successfully come up for the new arch, while existing ones will continue their lifecycle on the other arch resulting in a heterogeneous MachineSet.

Author: hmm...yeah that's a good point. i'll include it. but the question is should we allow a user to change existing machine set's instance type? or should this not be allowed going forward ?

Author: added this is in the risk/mitigations section ,but not sure what needs to be done for this. @JoelSpeed @elmiko any thoughts on this?

Reviewer 10: this is a great question @Prashanth684 , out team has been having some discussions recently about the types of validations we should be doing on the provider spec changes, and what we should allow/disallow.

changing the machine type/architecture sounds like something that could create headaches for our users. it's probably safer/cleaner to have them create new machinesets if they want to change something like that. but, with that said, i think our team should discuss this a little further so that we can provide a better answer. if you don't mind could you give us a few days to discuss this?

Author: no worries @elmiko .thanks for looking into this and do let me know what comes out of it.

Reviewer 10: our team talked about this a little bit today @Prashanth684 , and we plan on having a more in-depth discussion next tuesday at our architecture call. we have some questions about how this might be used and if we might investigate adopting MachineDeployments into the machine api for this kind of use case.

Author: updated the machine-api section to reflect the discussion we had today @elmiko . Thanks again!

Reviewer 10: np, thanks for joining us @Prashanth684

i had meant to come back here and leave a comment, apologies for the delay. we discussed this issue during the cloud team arch meeting today and i think the general consensus is that we prefer users to create new machinesets as opposed to updating the providerspec. i believe that if promote this style of user experience we should definitely create some more documentation about our reasons and also helpers for end users.

Reviewer 10: cc @JoelSpeed

> 

Reviewer 9: I think we need a Machine API Operator section right between Machine Config Operator and Hypershift

  - "@perdasilva - OLM changes"
  - "@mrunalp - MCO changes"
  - "@JoelSpeed - Machine api changes"
approvers:
  - "@bparees"
  - "@derekwaynecarr"
creation-date: 2022-01-17
last-updated: 2022-01-17
tracking-link:
  - https://issues.redhat.com/browse/RFE-2240
  - https://issues.redhat.com/browse/OCPPLAN-4577
status: implementable
---

# heterogeneous-architecture-cluster-support

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for tech preview and GA
- [ ] User-facing documentation is created in [Openshift-docs](https://github.com/Openshift/Openshift-docs/)

## Summary

This enhancement describes the plan to support provisioning and upgrades of heterogeneous architecture Openshift clusters.
A heterogeneous cluster is a cluster with nodes running on different architectures with a homogeneous control plane and worker node pools
supporting any combination of architectures (eg: x86 control plane + a mix of ARM/P/Z worker nodes). The node pools within the cluster are

> homogeneous. This document describes the support needed for heterogeneous architecture clusters in phases. There will be three major phases

Author: so as of today there is only a single machineconfig pool for all workers. there has not been a need for this so far as things like kubelet options have not had any architecture specific differences in terms of cmdline options. but there are different machinesets for x86 and ARM workers though.

> homogeneous. This document describes the support needed for heterogeneous architecture clusters in phases. There will be three major phases

Author: and also removed the sentence about node pools. that's was not correct. thanks.

- Release pipeline needs to be setup to construct manifestlist based release payloads.
- Upgrade support - CVO/Cincinnati support (heterogeneous -> heterogeneous payload upgrades only).
- Does not include support for:
  - image registry - internal image registry will not support manifest listed images.
  - imagestreams importing full manifestlist content (will continue to import a single image when referencing manifestlists).

Phase 2: Full basic functionality
- Imagestream manifestlist support.
- OLM support for node selection when deploying an operator and ecosystem improvements.
- Improve console's content filtering by making it architecture aware and passing that information to OLM.
- oc debug (and other commands) support for compute architecture.
- support toolbox.
- support for Power and Z architectures.
- Hypershift support for heterogeneous management+worker clusters.

Phase 3: Quality of life improvements
- First class install support - ability to create non-control-plane-architecture workers at install time and generate machineconfig
and machineset manifests for the same.
- Possibly reduce mirroring costs by looking into things like sparse manifestlist.

> - Upgrade support for homogeneous payload -> heterogeneous payload and vice versa.

Reviewer 3: the goal here is to allow a cluster admin to toggle their cluster back and forth between homogenous mode (smaller release payloads, but no heterogeneous compute) and heterogeneous mode (larger release payloads, but supports heterogeneous compute).

> - Upgrade support for homogeneous payload -> heterogeneous payload and vice versa.

Author: release payload. i'll make it clearer.

additional architectures if they were to use heterogeneous payloads. This means that in order for CVO to find the current release in the graph, it
will have to know about the existing homogeneous graph as well in addition to the new heterogeneous graph. There are few options on the best way
to make update recommendations especially in the case of upgrading from a homogeneous to a heterogeneous payload:

- Have the heterogeneous graph include the homogeneous releases so CVO can make recommendations based on the current homogeneous release image.
This is complicated by the fact that there will be multiple releases with the same version but different payloads. The trigger to update to a
heterogeneous payload would be an option provided by the administrator through either the ClusterVersion or Infrastructure API to "move" to the
heterogeneous graph.
- Have both the graphs be completely independent of each other and have an option in one of the aforementioned APIs to make the cluster heterogeneous
and CVO would request the heterogeneous graph, find the current release version and update to it.

> - Have an independent channel for heterogeneous payloads and have administrators switch channels if they want to move between homogeneous and

Reviewer 5: This is still an open question.

Author: yes so i have listed  the options here. in the "Open questions" section i have pointed back to this section as I have summarized the upgrade strategy overall.

Reviewer 2: What do you see as the risk that is being avoided?  Going from 4.11.0-x86 to 4.11.1-multi should be equivalent to going from 4.11.0-x86 to 4.11.1-x86, with the only difference being that the payload pullspecs will be manifestlists instead of manifests.

The cluster itself would only have x86 nodes regardless, so the actual image content the cluster consumes will be the same in the end.

Reviewer 11: I don't think the risk is high, I just don't want to be troubleshooting two concurrent changes in an uncommonly exercised migration. If we want to phase this in, offer it as a distinct migration step in one minor, then automate it in a later minor after having proven success in the field then I'm fine with that.

Author: @sdodson @LalatenduMohanty @wking have we reached  a consensus on the approach we will be taking here?

This is complicated by the fact that there will be multiple releases with the same version but different payloads. The trigger to update to a
heterogeneous payload would be an option provided by the administrator through either the ClusterVersion or Infrastructure API to "move" to the
heterogeneous graph.
- Have both the graphs be completely independent of each other and have an option in one of the aforementioned APIs to make the cluster heterogeneous
and CVO would request the heterogeneous graph, find the current release version and update to it.
- Have an independent channel for heterogeneous payloads and have administrators switch channels if they want to move between homogeneous and
heterogeneous payloads

These options are under discussion and a finalized proposal will be put forth in the ticket for upgrades referenced above along with the API that would
change to include the option to indicate a move to the heterogeneous payload. For a heterogeneous to a homogeneous payload upgrade, CVO would also need
to sanity check that there are no workers present which do not match the architecture of the homogeneous payload. If it does, it should fail the

> upgrade.

Author: Yes, this use case was seen as atypical. I'll discuss with Duncan about it, but i think we can say that a manual upgrade can be forced but this will not be officially supported.

> upgrade.

Author: added this to open questions.

be overriden with the manifest listed imagestream. Eventually the oc tooling would also be upgraded to point to the manifest listed image from
source for non control plane architecture workers.

### OLM

OLM would need to make sure that it deploys operators only to nodes whose architecture matches the architectures on which the operator is supported.
This would mean that the OLM would have to be told which architectures an operator supports by the Console operator. Today, OLM has a subscription
object which is used to tell it how to deploy an operator. It has support for the nodeselector field but this would not be sufficient to specify
filtering on multiple architectures. To this end, it would have to support the podAffinity field which uses the nodeselectorterms to filter on
multiple values.

Instead of having OLM dictate the operator deployment on specific nodes, the individual operators themselves would specify which architectures
they support through a nodeselector object which OLM will then provide the ability for the user to override if necessary. This information should

> also be added to the operands and  potentially be made configurable by the user requesting the operand.The operator-sdk framework could also be

Reviewer 12: Couple minor nits.

Extra space after `added to the operands and`
Space needed before `The operator-sdk`

running on.
- Console would provide a way for users to select which architecture nodes they would potentially want to deploy the operator on. This would be
filtered to provide a list of architectures that the operator supports.

### Install

For Phase 1, the way to add more workers to the cluster is to handcraft a machineset manifest file for a worker node and specify the architecture
specific details like instancetype, image references and deploy the machineset to get a worker node. This can be pseudo automated further by creating
the manifests using `openshift-install create manifests` and modifying or adding worker manifest files. Eventually the installer should support being
able to declare differing compute and control plane sections in the install-config mainly allowing architectures to be different. In doing so it would
also generate appropriate machineset and machineconfig files tailored to the specific architecture. The machineset file would contain the instancetype,
image id (AMI or Azure disk image etc..) and region corresponding to the architecture chosen for the workers.

### Open Questions

> - What would be the method to provide recommendation updates when moving from a homogeneous graph to a heterogeneous graph ?

Reviewer 11: Maybe alternatively have a conversion specific channel. They're to switch from stable-4.11 to heterogeneous-migration channel which has *only* 4.11.0-{x86_64,ppc64le,arm64,s390x} -> 4.11.0-multi edges. Once they've _upgraded_ to the corresponding mulit arch release they can then switch to multi-4.11.  CVO Implicitly knows if the release image is manifestlist then it should be fetching the multi arch graph?

> - What would be the method to provide recommendation updates when moving from a homogeneous graph to a heterogeneous graph ?

Author: i think this is in line with what Justin mentions in: https://github.com/openshift/enhancements/pull/1014#discussion_r791134030 , but there is no way back to homogeneous here? and as mentioned before we wouldn't want to support this scenario anyway.

> - What would be the method to provide recommendation updates when moving from a homogeneous graph to a heterogeneous graph ?

Author: added the suggestions to open questions.

> - What would be the method to provide recommendation updates when moving from a homogeneous graph to a heterogeneous graph ?

Author: resolved based on recent discussions and updated.

### Test Plan

Testing for Phase 1:
- QE would need to test heterogeneous nightly payloads before the entry is added to Cincinnati. This implies that the heterogeneous payloads will be
added to the candidate channel a bit later than the homogeneous payloads.
- QE would test heterogeneous payloads on Azure and AWS platforms with the cluster configuration mainly being x86 control plane and ARM worker nodes.
- Testing scenarios would include machine set scaling, upgrades involving ARM nodes and OVN testing on heterogeneous clusters.
- New release payload acceptance jobs on heterogeneous clusters.

Testing for Phase 2:
- With the addition of Power and Z architectures, there will be more permutations to test. We would need to figure out the absolute essential set of
permutations and the tests to run on those.
- Imagestream conformance tests.
- OLM and console testing
- Test various upgrade scenarios (homogeneous -> heterogeneous and vice versa)

### Graduation Criteria

> Phase 1 will have a pre-4.11 dev preview and 4.11 tech preview with documented limitations.

Reviewer 11: Do we set the techpreview no upgrade feature gate? In both cases?

Author: yes. updated to reflect this and removed the dev preview.

- oc debug (and other commands) support for compute architecture.
- support toolbox.
- support for Power and Z architectures.
- Hypershift support for heterogeneous management+worker clusters.

Phase 3: Quality of life improvements
- First class install support - ability to create non-control-plane-architecture workers at install time and generate machineconfig
and machineset manifests for the same.
- Possibly reduce mirroring costs by looking into things like sparse manifestlist.
- Upgrade support for homogeneous payload -> heterogeneous release payload and vice versa.

### User Stories

- As an enterprise customer looking to save compute resource cost, I want to be able to run part of my application workload on ARM instead of x86.
- As the owner of an application that runs partially on x86 and partially on ARM, I want to be able to run the entire application within a
single cluster so I do not need to pay the cost of higher latency between clusters, or worry about the HA-resilience of cross-cluster bridges.

### API Extensions

- Imagestreams API will introduce a new field to indicate that it will reference/import manifestlists. It would also contain a new child manifest

> field which would be for the image corresponding to the control plane architecture.

Reviewer 13: Can you include a golang like example of the changes you intend to make so that the API changes can be reviewed
eg
```
type ImageStreamsSpec struct {
  // Existing fields omitted for brevity

  // NewField is used to do, when it's not provided something happens, these are the boundaries and limitations of using the field
  // +kubebuilder:validation:...
  NewField []string `json:"newField"
```

Would be good to do this for both the changes mentioned in this section

Author: i think this is still in the works but @dmage should be able to provide something for the imagestreams and @wking can probably provide something for the ClusterVersion spec once they decide on the approach.

Author: i just came across this: https://github.com/openshift/enhancements/blob/master/enhancements/manifestlist/manifestlist-support.md#proposed-image-api-changes and i'm guessing the changes needed would be very similar.

- Upgrades to a heterogeneous release payload will require an additional field either in the Infrastructure or the ClusterVersion resource to indicate

> that the user wants to upgrade to a heterogeneous release payload.

Reviewer 11: @wking @LalatenduMohanty Can we quickly make a decision here if the CVO needs to know whether the cluster is homogeneous and thus fetches its native architecture update graph or if the CVO will infer from the fact that the release image is a manifestlist image that it's a heterogenous cluster?

I tend to think we should prefer being explicit here, so if for some reason all release images become manifestlist we're not scrambling to address this then.

Author: for upgrades, CVO would look at the newly added architecture field in the spec to figure out the type of the cluster. it is not clear how it would know on a fresh install of a cluster with heterogeneous payload - should it look at its own (CVO'S) image and if it is manifestlisted, assume the cluster is heterogeneous?

Reviewer 5: yeah, we not clear how CVO will be know if it is heterogeneous cluster from the pull spec. We will figure that out. But we are moving in that direction of adding an architecture field in ClusterVersion spec.

### Implementation Details/Notes/Constraints

Operators supporting multiple architectures cannot stop supporting a given architecture since it would break heterogeneous clusters which are
trying to upgrade.

### Risks and Mitigations

Heterogeneous release payloads will only ship if the corresponding homogeneous release payloads for all architectures can ship and are not broken due

> to a bug in a specific architecture.

Author: as of today the homogeneous payloads for all arches ship together. what this means is - if there is a blocker on one architecture, other arches would be blocked as well.

> to a bug in a specific architecture.

Reviewer 11: @Prashanth684 that detail makes me think this isn't really a risk but I'm also fine with leaving it in if we add that detail. Essentially that's the mitigation "But we wouldn't ship homogeneous releases in those cases either." or something.

## Design Details

### RHCOS Support for ARM Azure

The RHCOS pipeline will need to generate Azure disk images for ARM. However as of today, RHEL does not support hyper-v for ARM and that support is
landing in RHEL8.6. Hence 4.11 builds of RHCOS for ARM Azure will be based on RHEL8.6. It is TBD whether a separate pipeline with early access RHCOS

> builds will be needed as RHEL8.6 would not have GA'ed before 4.11.

Reviewer 8: I think we can get 8.6 RHCOS out pretty quickly.

> builds will be needed as RHEL8.6 would not have GA'ed before 4.11.

Reviewer 2: Intent was to get phase 0 into some user hands early to play with and the primary audience for that needs it on azure. So we can do pre testing elsewhere, but in terms of a deliverable phase 0, to be useful it needs azure support.

> builds will be needed as RHEL8.6 would not have GA'ed before 4.11.

Author: we were able to spin up an rhcos vhd with RHEL 8.6 content and test it out on the Azure ARM test instances provided by adding ARM worker nodes to an x86 cluster and it works fine so afaik nothing special.

> builds will be needed as RHEL8.6 would not have GA'ed before 4.11.

Author: changed the wording to make it simpler.

also be added to the operands and potentially be made configurable by the user requesting the operand. The operator-sdk framework could also be
potentially leveraged to support the generation of this supported architecture list when creating operators.

### Machine Config Operator

No changes are needed in the MCO for Phase 1 as the machine-os-content image would be a manifest listed image and the machine config daemon would
extract the relevant architecture's machine-os-content based on the node it runs on. In the future, architecture specific configurations should
not be required, instead they need to be templatized and be generalized for certain items, for example kubelet system reserved memory could scale
based on page size.

### Machine API

- As of today, machinesets can be mutated to change the instance type and image id and this could potentially result in inconsistency between existing
machinesets and newly created machinesets. If the instance type and image is changed to one of a different architecture, there would be workers of
different architectures with the same machineset definition. While this may not be a problem, it doesn't make for a good user experience. Hence,

> machineset specs will not allowed to be mutated. Anytime a change is required, a new machineset needs to be created and the old one needs to be scaled

Reviewer 14: What happens today in a homogeneous cluster if the instance type is modified to one that uses a different architecture? Do we already block users from editing that field or is this a big API change?

Author: today we allow users to change instance types. what happens is - the  newer machines created will be of the new instance type while the older machines will persist, ending up in a mismatch. upon discussion with the cloud team the idea was to block this change and ask users to create a new machineset instead of changing existing ones.

Reviewer 13: Is there an easy way that we can do this across platforms? While we have admission webhooks already for some platforms, changing the behaviour to make part of the MachineSet providerSpec immutable where it wasn't before sounds like a breaking change.

Are we going to need to enumerate amd64 and arm64 instance types and make sure that if your MachineSet was created as one, it cannot be updated to the other? 

Note, in our existing webhooks, we don't have the old spec when validating, so that will need to be changed

Have we weighed up the complexity of making this immutable vs the user shooting themselves in the foot factor? Do we need a technical solution for this?

Author: yes part of the problem today is that there is no way to differentiate arches today. i remember in the discussion, it was agreed upon to make it immutable, but not sure the details were hashed out after that. @elmiko might have further ideas.

Author: i would assume this would mean making instance type immutable, but not sure if that would apply to the spec as a whole to make it immutable in it's entirety.

Reviewer 13: We already today have instructions in our product docs (those for resizing the control plane) that talk about changing the instance type of a Machine when you resize it out of band, we need to make sure we don't break scenarios like this that are already documented.

Reviewer 10: reading back here, i feel like if we decide to start doing validations on the fields of the provider spec we will get into some serious complexity quickly. making them immutable will probably shock some users who are accustomed to having the ability to change the provider spec, but imo will lead us to a better place in terms of best cluster practices. from a certain standpoint creating a new machineset and updating the provider spec on a current one aren't much different. the user will still need to scale up new instances to get the new provider changes.

Author: have we decided whether this solution is worth pursuing or leaving as is? @JoelSpeed @elmiko ?

down.
- Changes are needed in the autoscaler to make it aware of the node architecture when autoscaling from zero. The `kubernetes-io-arch` label needs to be
set when building the NodeInfo object so the autoscaler can pick the correct pool to scale up for allocating nodes of a particular architecture.

### Hypershift

The addition of heterogeneous support in Openshift also opens up a new dimension for [support][OCPPLAN-5684] in Hypershift where the control plane 
and the workers could run on different architectures. There would be certain changes needed to realize this as there are limitations which need to
be addressed:

- When creating the hosted cluster, the latest x86 release payload is picked by default. This would have to change to pick a heterogeneous payload

> when creating the hosted cluster on mixed architecture workers.

Reviewer 14: Why wouldn't the selection be based on the channel, as described above in the CVO section?

Author: are you suggesting that hypershift should read the pullsepc from the clusterversion object ? today the hosted control plane image defaults to picking an x86 image: https://github.com/openshift/hypershift/blob/main/cmd/version/version.go#L17

Author: to be clear, the hosted cluster creation uses the management cluster's payload -  so if the management cluster is heterogeneous, we don't have a problem. But if the management cluster is of a single arch , we cannot create a nodepool/hosted cluster of a different arch without tweaks

Reviewer 2: > to be clear, the hosted cluster creation uses the management cluster's payload 

no it doesn't.  When creating a guest cluster you pick the payload to use.  Hypershift host/management clusters can install guest/hosted clusters of varying versions, not just the same version as the host/management cluster.

Author: that's correct. i meant that by default the management cluster's spec is used. The bigger problem here is - the hosted cluster's ignition server uses the nodepool's mco to generate ignition for the nodepools. Likewise, for some of the components deployed on the nodepool, the hosted cluster's image pullspecs are used. so even if we did have a management cluster and hosted cluster on a single arch, we cannot have a nodepool of a different arch without using a heterogeneous payload.

  to be imagestreams added for individual architectures and they would be picked based on the node architecture.
  - references to some images like cluster-api-provider-aws, apiserver-network-proxy etc.. are x86. these would need to change to provide the images
  for other architectures too. These images can be overridden through annotations today.
- The nodepool object defaults to picking an x86 instancetype and AMI. The nodepool API does expose these fields so they can be changed manually to
architecture appropriate values. The API has to make a provision to include the architecture value so the AMI and instance type can be chosen
accordingly.

### Console

- As of today console filters content based on control plane architecture. This needs to be entirely removed or made configurable for
heterogeneous clusters. The operator should be able to filter out the set of OLM operators that, if installed will be able to find
worker nodes to run on. This means that it has to display only operators that can be supported by architectures that the workers are

> running on.

Reviewer 14: Are we sure there are no OLM operators that expect to be able to run on the control plane hosts?

Author: there are operators which deploy on control plane as well. so yes in that case the supported arches will be cp arch + worker arch

The above behavior is followed for homogeneous architecture payloads today and it is expected that oc would support it for heterogeneous payloads too.
The only difference for heterogeneous payloads though is that, the extract command would return a set of tools for all supported payload architectures
when the `--command-os=*` flag is used. The table below describes the combination of os/architectures for which the oc client and the installer binary
would be extracted.

| Client platform |	Image arch | Image | Path | Tarball |
| :--- | :--- | :--- | :--- | :--- |
| linux/amd64 |	amd64 |	cli | /usr/bin/oc |	openshift-client-linux-amd64-%s.tar.gz |
| linux/arm64	| arm64	| cli	| /usr/bin/oc | openshift-client-linux-arm64-%s.tar.gz |
| linux/ppc64le | ppc64le | cli | /usr/bin/oc | openshift-client-linux-ppc64le-%s.tar.gz |
| linux/s390x | s390x | cli | /usr/bin/oc | openshift-client-linux-s390x-%s.tar.gz |
| darwin/amd64 | amd64* | cli-artifacts | /usr/share/openshift/mac/oc | openshift-client-mac-%s.tar.gz |
| darwin/arm64 | arm64* |	cli-artifacts |	/usr/share/openshift/mac_arm64/oc | openshift-client-mac-arm64-%s.tar.gz |

> | windows/amd64 |	amd64* | cli-artifacts | /usr/share/openshift/windows/oc.exe | openshift-client-windows-%s.tar.gz |

Reviewer 15: Should we add windows/arm64

Reviewer 15: its probably offtopic here, but should be asked in a more appropriate forum

| Install platform | Image arch | Image |	Path | Tarball |
| :--- | :--- | :--- | :--- | :--- |
| linux/amd64 |	amd64 | installer |	/usr/bin/openshift-install | openshift-install-linux-amd64-%s.tar.gz |
| linux/arm64	| arm64	| installer | /usr/bin/openshift-install | openshift-install-linux-arm64-%s.tar.gz |
| linux/ppc64le |	ppc64le | installer | /usr/bin/openshift-install | openshift-install-linux-ppc64le-%s.tar.gz |
| linux/s390x	| s390x	| installer	| /usr/bin/openshift-install | openshift-install-linux-s390x-%s.tar.gz |
| darwin/amd64 | amd64* | installer-artifacts |	/usr/share/openshift/mac/openshift-install | openshift-install-mac-%s.tar.gz |

> | darwin/arm64 | arm64* | installer-artifacts | /usr/share/openshift/mac_arm64/openshift-install | openshift-install-mac-arm64-%s.tar.gz |

Reviewer 16: A side effect of having a single installer for all payload architectures means that we can no longer rely on hardcoding the payload architectures at build time (as implemented in https://github.com/openshift/installer/pull/4870 and https://github.com/openshift/installer/pull/5107).  Instead, `compute.architecture` and `controlPlane.architecture` will need to be mandatory with no default, and "openshift-install create install-config" modified to add a question to select an architecture among those supported by the selected platform (e.g. `aws` => `['amd64','arm64']`).

Author: hmm..yes true..i need to add this as well..thanks for catching this @yselkowitz . also related is - how would the `oc extract` command work? would it extract oc/installer binaries from each of the manifest images in the manifestlisted cli/installer artifacts image? the extract would also have to embed the heterogeneous release payload pullspec in the installer binaries.

Today, OLM has a subscription object which is used to tell it how to deploy an operator. It has support for the nodeselector field but this would
not be sufficient to specify filtering on multiple architectures. To this end, it would have to support the podAffinity field which uses the
nodeselectorterms to filter on multiple values.

Instead of having OLM dictate the operator deployment on specific nodes, the individual operators themselves would specify which architectures
they support through a nodeselector object which OLM will then provide the ability for the user to override if necessary. This information should
also be added to the operands and potentially be made configurable by the user requesting the operand. The operator-sdk framework could also be
potentially leveraged to support the generation of this supported architecture list when creating operators.

### Machine Config Operator

No changes are needed in the MCO for Phase 1 as the machine-os-content image would be a manifest listed image and the machine config daemon would

> extract the relevant architecture's machine-os-content based on the node it runs on. In the future, architecture specific configurations should

Reviewer 17: I'm not following this. Doesn't this presuppose the MCO is already running on the target architecture you're trying to pull configs for?

Author: it does. I have just called out the fact that when MCO pivots to the machine-os-content specified by the payload, it does install the relevant architecture specific machine-os from the manifestlisted machine-os image (as it uses oc image extract which uses the docker APIs)
